File: docx-parser-converter-js/src/parsers/tables/table-properties-parser.ts
/**
 * Table properties parser for DOCX documents
 * Parses table properties from w:tblPr elements
 */

import type { 
  TableProperties, 
  TableWidth, 
  TableIndent, 
  TableLook, 
  TableCellBorders, 
  ShadingProperties, 
  MarginProperties 
} from '@/models/table-models.js';
import { 
  TablePropertiesModel, 
  TableWidthModel, 
  TableIndentModel, 
  TableLookModel, 
  TableCellBordersModel, 
  ShadingPropertiesModel, 
  MarginPropertiesModel,
  BorderPropertiesModel 
} from '@/models/table-models.js';
import { convertTwipsToPoints } from '@/utils/unit-conversion.js';
import { getFirstChildElement } from '@/utils/xml-utils.js';
import { BaseParser } from '../base-parser.js';

/**
 * Table properties parser class
 */
export class TablePropertiesParser extends BaseParser<TableProperties> {
  constructor(options: Record<string, unknown> = {}) {
    super('TablePropertiesParser', options);
  }

  /**
   * Parse XML object into TableProperties model
   * @param xmlObj - Parsed XML object containing w:tblPr or w:tblPrEx element
   * @returns Promise resolving to TableProperties model
   */
  protected async parseInternal(xmlObj: Record<string, unknown>): Promise<TableProperties> {
    // Extract w:tblPr or w:tblPrEx element
    let tblPrElement: Record<string, unknown>;

    // Try standard format first
    if (xmlObj['w:tblPr']) {
      const tblPrValue = xmlObj['w:tblPr'];
      
      // Handle case where w:tblPr is an array
      if (Array.isArray(tblPrValue)) {
        if (tblPrValue.length === 0) {
          throw new Error('Empty w:tblPr array found in XML');
        }
        tblPrElement = tblPrValue[0] as Record<string, unknown>;
      } else {
        tblPrElement = tblPrValue as Record<string, unknown>;
      }
    } 
    // Try extended format
    else if (xmlObj['w:tblPrEx']) {
      const tblPrExValue = xmlObj['w:tblPrEx'];
      
      // Handle case where w:tblPrEx is an array
      if (Array.isArray(tblPrExValue)) {
        if (tblPrExValue.length === 0) {
          throw new Error('Empty w:tblPrEx array found in XML');
        }
        tblPrElement = tblPrExValue[0] as Record<string, unknown>;
      } else {
        tblPrElement = tblPrExValue as Record<string, unknown>;
      }
    } 
    else if (xmlObj && typeof xmlObj === 'object' && Object.keys(xmlObj).length > 0) {
      // If the root object itself is the w:tblPr or w:tblPrEx element
      tblPrElement = xmlObj;
    } else {
      throw new Error('No w:tblPr or w:tblPrEx element found in XML');
    }

    return this.parseTablePropertiesElement(tblPrElement);
  }

  /**
   * Parse table properties element
   * @param tblPrElement - w:tblPr element
   * @returns Parsed TableProperties
   */
  private async parseTablePropertiesElement(tblPrElement: Record<string, unknown>): Promise<TableProperties> {
    const props: Partial<TableProperties> = {};

    // Parse table style (w:tblStyle)
    props.tblStyle = this.parseTableStyle(tblPrElement);
    
    // Parse table width (w:tblW)
    props.tblW = await this.parseTableWidth(tblPrElement);
    
    // Parse justification (w:jc)
    props.justification = this.parseJustification(tblPrElement);
    
    // Parse table indent (w:tblInd)
    props.tblInd = await this.parseTableIndent(tblPrElement);
    
    // Parse table cell margins (w:tblCellMar)
    props.tblCellMar = await this.parseTableCellMargins(tblPrElement);
    
    // Parse table borders (w:tblBorders)
    props.tblBorders = await this.parseTableBorders(tblPrElement);
    
    // Parse shading (w:shd)
    props.shd = await this.parseShading(tblPrElement);
    
    // Parse table layout (w:tblLayout)
    props.tblLayout = this.parseTableLayout(tblPrElement);
    
    // Parse table look (w:tblLook)
    props.tblLook = await this.parseTableLook(tblPrElement);

    // Set convenience properties to match test expectations
    if (props.tblW?.width) {
      props.width = props.tblW.width;
    }
    if (props.justification) {
      props.alignment = props.justification;
    }
    if (props.tblBorders) {
      props.borders = props.tblBorders;
    }
    if (props.tblLook) {
      props.look = props.tblLook;
    }
    if (props.tblStyle) {
      props.style = props.tblStyle;
    }
    if (props.tblInd?.width) {
      props.indent = props.tblInd.width;
    }

    return TablePropertiesModel.create(props);
  }

  /**
   * Parse table style from w:tblStyle element
   * @param tblPrElement - w:tblPr element
   * @returns Table style or undefined
   */
  private parseTableStyle(tblPrElement: Record<string, unknown>): string | undefined {
    const tblStyle = getFirstChildElement(tblPrElement, 'w:tblStyle');
    return tblStyle ? this.getAttribute(tblStyle, 'w:val') : undefined;
  }

  /**
   * Parse table width from w:tblW element
   * @param tblPrElement - w:tblPr element
   * @returns TableWidth or undefined
   */
  private async parseTableWidth(tblPrElement: Record<string, unknown>): Promise<TableWidth | undefined> {
    const tblW = getFirstChildElement(tblPrElement, 'w:tblW');
    if (!tblW) return undefined;

    const type = this.getAttribute(tblW, 'w:type');
    const widthValue = this.getAttribute(tblW, 'w:w');
    
    let width: number | undefined;
    if (widthValue) {
      const parsedWidth = parseInt(widthValue, 10);
      if (!isNaN(parsedWidth)) {
        // Convert from twips to points if type is dxa
        width = type === 'dxa' ? convertTwipsToPoints(parsedWidth) : parsedWidth;
      }
    }

    return TableWidthModel.create({
      type,
      width,
    });
  }

  /**
   * Parse justification from w:jc element
   * @param tblPrElement - w:tblPr element
   * @returns Justification value or undefined
   */
  private parseJustification(tblPrElement: Record<string, unknown>): string | undefined {
    const jc = getFirstChildElement(tblPrElement, 'w:jc');
    return jc ? this.getAttribute(jc, 'w:val') : undefined;
  }

  /**
   * Parse table indent from w:tblInd element
   * @param tblPrElement - w:tblPr element
   * @returns TableIndent or undefined
   */
  private async parseTableIndent(tblPrElement: Record<string, unknown>): Promise<TableIndent | undefined> {
    const tblInd = getFirstChildElement(tblPrElement, 'w:tblInd');
    if (!tblInd) return undefined;

    const type = this.getAttribute(tblInd, 'w:type');
    const widthValue = this.getAttribute(tblInd, 'w:w');
    
    let width: number | undefined;
    if (widthValue) {
      const parsedWidth = parseInt(widthValue, 10);
      if (!isNaN(parsedWidth)) {
        // Convert from twips to points if type is dxa
        width = type === 'dxa' ? convertTwipsToPoints(parsedWidth) : parsedWidth;
      }
    }

    return TableIndentModel.create({
      type,
      width,
    });
  }

  /**
   * Parse table cell margins from w:tblCellMar element
   * @param tblPrElement - w:tblPr element
   * @returns MarginProperties or undefined
   */
  private async parseTableCellMargins(tblPrElement: Record<string, unknown>): Promise<MarginProperties | undefined> {
    const tblCellMar = getFirstChildElement(tblPrElement, 'w:tblCellMar');
    if (!tblCellMar) return undefined;

    const margins: Partial<MarginProperties> = {};

    // Parse each margin direction, including start/end as fallbacks for left/right
    const marginMappings = [
      { prop: 'top', element: 'w:top' },
      { prop: 'bottom', element: 'w:bottom' },
      { prop: 'left', element: 'w:left' },
      { prop: 'right', element: 'w:right' },
    ];

    for (const { prop, element } of marginMappings) {
      let marginElement = getFirstChildElement(tblCellMar, element);
      
      // Fallback for left/right to start/end
      if (!marginElement && prop === 'left') {
        marginElement = getFirstChildElement(tblCellMar, 'w:start');
      } else if (!marginElement && prop === 'right') {
        marginElement = getFirstChildElement(tblCellMar, 'w:end');
      }

      if (marginElement) {
        const widthValue = this.getAttribute(marginElement, 'w:w');
        if (widthValue) {
          const parsedWidth = parseInt(widthValue, 10);
          if (!isNaN(parsedWidth)) {
            margins[prop as keyof MarginProperties] = convertTwipsToPoints(parsedWidth);
          }
        }
      }
    }

    return Object.keys(margins).length > 0 ? MarginPropertiesModel.create(margins) : undefined;
  }

  /**
   * Parse table borders from w:tblBorders element
   * @param tblPrElement - w:tblPr element
   * @returns TableCellBorders or undefined
   */
  private async parseTableBorders(tblPrElement: Record<string, unknown>): Promise<TableCellBorders | undefined> {
    const tblBorders = getFirstChildElement(tblPrElement, 'w:tblBorders');
    if (!tblBorders) return undefined;

    const borders: Partial<TableCellBorders> = {};

    // Parse each border direction
    for (const direction of ['top', 'left', 'bottom', 'right', 'insideH', 'insideV']) {
      const borderElement = getFirstChildElement(tblBorders, `w:${direction}`);
      if (borderElement) {
        const color = this.getAttribute(borderElement, 'w:color');
        const sizeValue = this.getAttribute(borderElement, 'w:sz');
        const spaceValue = this.getAttribute(borderElement, 'w:space');
        const val = this.getAttribute(borderElement, 'w:val');

        const size = sizeValue ? parseInt(sizeValue, 10) : undefined;
        const space = spaceValue ? parseInt(spaceValue, 10) : undefined;

        borders[direction as keyof TableCellBorders] = BorderPropertiesModel.create({
          color: color && color !== 'auto' ? `#${color}` : color,
          size: !isNaN(size as number) ? size : undefined,
          space: !isNaN(space as number) ? space : undefined,
          val,
        });
      }
    }

    return Object.keys(borders).length > 0 ? TableCellBordersModel.create(borders) : undefined;
  }

  /**
   * Parse shading from w:shd element
   * @param tblPrElement - w:tblPr element
   * @returns ShadingProperties or undefined
   */
  private async parseShading(tblPrElement: Record<string, unknown>): Promise<ShadingProperties | undefined> {
    const shd = getFirstChildElement(tblPrElement, 'w:shd');
    if (!shd) return undefined;

    const fill = this.getAttribute(shd, 'w:fill');
    const val = this.getAttribute(shd, 'w:val');
    const color = this.getAttribute(shd, 'w:color');

    return ShadingPropertiesModel.create({
      fill: fill && fill !== 'auto' ? `#${fill}` : fill,
      val,
      color: color && color !== 'auto' ? `#${color}` : color,
    });
  }

  /**
   * Parse table layout from w:tblLayout element
   * @param tblPrElement - w:tblPr element
   * @returns Table layout or undefined
   */
  private parseTableLayout(tblPrElement: Record<string, unknown>): string | undefined {
    const tblLayout = getFirstChildElement(tblPrElement, 'w:tblLayout');
    return tblLayout ? this.getAttribute(tblLayout, 'w:type') : undefined;
  }

  /**
   * Parse table look from w:tblLook element
   * @param tblPrElement - w:tblPr element
   * @returns TableLook or undefined
   */
  private async parseTableLook(tblPrElement: Record<string, unknown>): Promise<TableLook | undefined> {
    const tblLook = getFirstChildElement(tblPrElement, 'w:tblLook');
    if (!tblLook) return undefined;

    const val = this.getAttribute(tblLook, 'w:val');
    const firstRow = this.getBooleanAttribute(tblLook, 'w:firstRow');
    const lastRow = this.getBooleanAttribute(tblLook, 'w:lastRow');
    const firstColumn = this.getBooleanAttribute(tblLook, 'w:firstColumn');
    const lastColumn = this.getBooleanAttribute(tblLook, 'w:lastColumn');
    const noHBand = this.getBooleanAttribute(tblLook, 'w:noHBand');
    const noVBand = this.getBooleanAttribute(tblLook, 'w:noVBand');

    return TableLookModel.create({
      val,
      firstRow,
      lastRow,
      firstColumn,
      lastColumn,
      noHBand,
      noVBand,
    });
  }
} 

File: docx_parser_converter_python/parsers/document/table_properties_parser.py
"""Parser for table properties elements."""

from lxml.etree import _Element as Element

from models.document.table import TableLook, TableProperties
from parsers.common.border_parser import parse_table_borders
from parsers.common.shading_parser import parse_shading
from parsers.common.width_parser import parse_width
from parsers.document.table_cell_properties_parser import parse_table_cell_margins
from parsers.utils import find_child, get_attribute


def parse_table_look(element: Element | None) -> TableLook | None:
    """Parse <w:tblLook> element.

    XML Example:
        <w:tblLook w:firstRow="1" w:lastRow="0" w:firstColumn="1"
                   w:lastColumn="0" w:noHBand="0" w:noVBand="1"/>

    Args:
        element: The <w:tblLook> element or None

    Returns:
        TableLook model or None if element is None
    """
    if element is None:
        return None

    # These can be "0", "1", "true", "false" etc.
    first_row = get_attribute(element, "firstRow")
    last_row = get_attribute(element, "lastRow")
    first_column = get_attribute(element, "firstColumn")
    last_column = get_attribute(element, "lastColumn")
    no_h_band = get_attribute(element, "noHBand")
    no_v_band = get_attribute(element, "noVBand")

    def to_bool(val: str | None) -> bool | None:
        if val is None:
            return None
        return val.lower() not in ("0", "false", "off")

    return TableLook(
        first_row=to_bool(first_row),
        last_row=to_bool(last_row),
        first_column=to_bool(first_column),
        last_column=to_bool(last_column),
        no_h_band=to_bool(no_h_band),
        no_v_band=to_bool(no_v_band),
    )


def parse_table_properties(element: Element | None) -> TableProperties | None:
    """Parse <w:tblPr> element.

    XML Example:
        <w:tblPr>
            <w:tblStyle w:val="TableGrid"/>
            <w:tblW w:w="5000" w:type="pct"/>
            <w:jc w:val="center"/>
            <w:tblBorders>...</w:tblBorders>
        </w:tblPr>

    Args:
        element: The <w:tblPr> element or None

    Returns:
        TableProperties model or None if element is None
    """
    if element is None:
        return None

    # Table style reference
    tbl_style_elem = find_child(element, "tblStyle")
    tbl_style = get_attribute(tbl_style_elem, "val") if tbl_style_elem is not None else None

    # Table width
    tbl_w = parse_width(find_child(element, "tblW"))

    # Table justification
    jc_elem = find_child(element, "jc")
    jc = get_attribute(jc_elem, "val") if jc_elem is not None else None

    # Table indentation
    tbl_ind = parse_width(find_child(element, "tblInd"))

    # Table borders
    tbl_borders = parse_table_borders(find_child(element, "tblBorders"))

    # Table shading
    shd = parse_shading(find_child(element, "shd"))

    # Table layout (fixed or autofit)
    tbl_layout_elem = find_child(element, "tblLayout")
    tbl_layout = get_attribute(tbl_layout_elem, "type") if tbl_layout_elem is not None else None

    # Default cell margins
    tbl_cell_mar = parse_table_cell_margins(find_child(element, "tblCellMar"))

    # Table look (conditional formatting flags)
    tbl_look = parse_table_look(find_child(element, "tblLook"))

    # Table caption
    tbl_caption_elem = find_child(element, "tblCaption")
    tbl_caption = get_attribute(tbl_caption_elem, "val") if tbl_caption_elem is not None else None

    # Table description
    tbl_description_elem = find_child(element, "tblDescription")
    tbl_description = (
        get_attribute(tbl_description_elem, "val") if tbl_description_elem is not None else None
    )

    return TableProperties(
        tbl_style=tbl_style,
        tbl_w=tbl_w,
        jc=jc,
        tbl_ind=tbl_ind,
        tbl_borders=tbl_borders,
        shd=shd,
        tbl_layout=tbl_layout,
        tbl_cell_mar=tbl_cell_mar,
        tbl_look=tbl_look,
        tbl_caption=tbl_caption,
        tbl_description=tbl_description,
    )
